src/compiler/sys.ts

```diff
@@ -418,50 +418,41 @@ namespace ts {
         }
     }
 
-    function createSingleFileWatcherPerName(
-        watchFile: HostWatchFile,
-        useCaseSensitiveFileNames: boolean
-    ): HostWatchFile {
-        interface SingleFileWatcher {
-            watcher: FileWatcher;
-            refCount: number;
-        }
-        const cache = new Map<string, SingleFileWatcher>();
-        const callbacksCache = createMultiMap<FileWatcherCallback>();
+    interface SingleFileWatcher<T extends FileWatcherCallback | FsWatchCallback>{
+        watcher: FileWatcher;
+        callbacks: T[];
+    }
+    function createSingleWatcherPerName<T extends FileWatcherCallback | FsWatchCallback>(
+        cache: Map<SingleFileWatcher<T>>,
+        useCaseSensitiveFileNames: boolean,
+        name: string,
+        callback: T,
+        createWatcher: (callback: T) => FileWatcher,
+    ): FileWatcher {
         const toCanonicalFileName = createGetCanonicalFileName(useCaseSensitiveFileNames);
+        const path = toCanonicalFileName(name);
+        const existing = cache.get(path);
+        if (existing) {
+            existing.callbacks.push(callback);
+        }
+        else {
+            cache.set(path, {
+                watcher: createWatcher((
+                    // Cant infer types correctly so lets satisfy checker
+                    (param1: any, param2: never, param3: any) => cache.get(path)?.callbacks.slice().forEach(cb => cb(param1, param2, param3))
+                ) as T),
+                callbacks: [callback]
+            });
+        }
 
-        return (fileName, callback, pollingInterval, options) => {
-            const path = toCanonicalFileName(fileName);
-            const existing = cache.get(path);
-            if (existing) {
-                existing.refCount++;
-            }
-            else {
-                cache.set(path, {
-                    watcher: watchFile(
-                        fileName,
-                        (fileName, eventKind, modifiedTime) => forEach(
-                            callbacksCache.get(path),
-                            cb => cb(fileName, eventKind, modifiedTime)
-                        ),
-                        pollingInterval,
-                        options
-                    ),
-                    refCount: 1
-                });
+        return {
+            close: () => {
+                const watcher = cache.get(path);
+                if (!watcher) return;

```

Comment:
Can this happen?  Why would it not be in the cache?

Reply:
Same as before.. defensive to handle the exceptions that normally result from unrelated code